### PR TITLE
Returning updated swaggerObject

### PR DESCRIPTION
**Pull request by pktippa**
The pull request is fix for https://github.com/strongloop/loopback-component-explorer/issues/140

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by pktippa** [comment]
@bajtos can you please accept this pull request?

**Comment by bajtos** [comment]
Hi @pktippa, thank you for the pull request. To be honest, I don't see how moving a variable declaration (not initialisation!) to a higher scope solves the problem.

Please add a unit-test that fails with the current implementation and passes with your fix in place - that's the best way how to show the correctness of the proposed solution. You can take a look at the unit-test in #22 for inspiration, just note that it has to be updated quite a lot to fit into the current codebase.

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by bajtos** [comment]
@pktippa I am not sure how familiar you are with GitHub pull request. When reworking this patch, please keep using the same git branch & github pull request, do not open a new one. You can simply keep adding more commits here, as we can clean the history at the end, just before merging the code.

**Comment by pktippa** [comment]
@bajtos I am not that familiar with writing test cases, please help me.

**Comment by bajtos** [comment]
@pktippa Start with reading the unit-test in #22, it's a good one. Rework it to work in the current codebase. Then think about any edge cases that would not be covered by that test and write additional tests for these edge cases.

When adding a new test, it's important to see it fail for the reason you expected. This has multiple purposes: it ensures the test actually tests something (as opposed to tests that always passes because they are based on incorrect assumptions) and also that the failure message is helpful enough to identify the source of the problem.

**Comment by pktippa** [comment]
@bajtos The issue is with [express](https://github.com/strongloop/express) itself. I have created a sample express app using express [generator](https://github.com/expressjs/generator) and added routes to create dynamic routes and respond with global and local scope varaibles, added test cases written in [supertest](https://github.com/visionmedia/supertest) and published the project in github named [express-route-bug](https://github.com/pk-nodejs/express-route-bug), please fork/check out the project run "npm test" to validate it.

If you find its a valid bug, Please report the same to express [issues](https://github.com/strongloop/express/issues).

I think since the bug is with express itself, Let me know if I still have to write test cases for this issue #140 .

**Comment by bajtos** [comment]
@pktippa I am afraid I don't see how is your test case in express-route-bug related to the work made here?

Let me rephrase the test scenario from #22.

1) Create a loopback app
2) Setup explorer on this app
3) `GET /explorer/swagger/json` to ensure any lazy initialisation is executed
3) Add a new model
4) `GET /explorer/swagger/json` and check that the new model is included

**Comment by pktippa** [comment]
@bajtos I have completed writing test cases for new model creation and existence of new model in the swagger.json

**Comment by pktippa** [comment]
@bajtos for your query.
>To be honest, I don't see how moving a variable declaration (not initialisation!) to a higher scope solves the problem.


Please go through the conversation below and explanation given by @dougwilson
https://github.com/strongloop/express/issues/2822

**Comment by pktippa** [comment]
Hi @bajtos can you please look into the test cases and accept this PR.

File: test/explorer.test.js [hunk 1]
```diff
@@ -237,4 +264,14 @@ describe('explorer', function() {
...
   }
+
+  function createNewModel(app, explorerBase,modelName) {
+    var Model = loopback.PersistedModel.extend(modelName);
+    Model.attachTo(loopback.memory());
+    app.model(Model);
+
+    explorer(app, { mountPath: explorerBase });
+    app.set('legacyExplorer', false);
+    app.use(app.get('restApiRoot') || '/', loopback.rest());
```
Comment (bajtos) [inline test/explorer.test.js]: You should not need to call `explorer(app)` again, lines 274-275 should be removed as well.

When you call `explorer(app)` again, it will add a duplicate set of routes to the application. Each express route incurs a small performance penalty, therefore adding unnecessary duplicate routes can cause performance degradation.

Also the code adding new models should not need to be aware of the fact that the app is using explorer. Your current solution creates a coupling between unrelated parts of the system which is a bad practice.

So, please remove L273-L275, I am suggesting to move L269-L271 back to the test method and get rid of `createNewModel` entirely.

File: index.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ var cors = require('cors');
 var createSwaggerObject = require('loopback-swagger').generateSwaggerSpec;
 var SWAGGER_UI_ROOT = require('strong-swagger-ui/index').dist;
 var STATIC_ROOT = path.join(__dirname, 'public');
+var swaggerObject;
```
Comment (bajtos) [inline index.js]: This creates a static global singleton shared by all applications. Consider the following example:

```js
var app1 = loopback();
// setup models, etc.
explorer(app1);

var app2 = loopback();
// setup models, etc.
explorer(app2);
```

With the solution proposed in this pull request, `app1` will serve explorer describing `app2`, which is clearly a bug.

The trick is to keep `swaggerObject` variable defined inside `mountSwagger` and listen for `modelRemoted` event to update it whenever a new model is added.


```js
function mountSwagger(loopbackApplication, swaggerApp, opts) {
  var  swaggerObject = createSwaggerObject(loopbackApplication, opts);

  loopbackApplication.on('modelRemoted', function() {
    swaggerObject = createSwaggerObject(loopbackApplication, opts);
  });

  var resourcePath = opts && opts.resourcePath || 'swagger.json';
  // etc.
}
```

**Comment by pktippa** [comment]
Hi @bajtos 

I have modified the code and tested, please check it once.

Please let me know revert commits and push a single commit for this pull request and let me know the steps as well, since I am new to git and github.

**Comment by bajtos** [comment]
Hi @pktippa, thank you for the update, the new code looks much better. Unfortunately you have messed up whitespace in your last commit, see [diff](https://github.com/strongloop/loopback-component-explorer/pull/141/files) and [diff ignoring whitespace changes](https://github.com/strongloop/loopback-component-explorer/pull/141/files?w=1). Could you please restore the original whitespace? It will make it much easier for me to comment on your changes.

> Please let me know revert commits and push a single commit for this pull request and let me know the steps as well, since I am new to git and github.

You need to run `git rebase -i master`, then choose "squash" for all commits except the first (top) one. Once rebased, you need `git push -f`.

Don't worry about it now, it's best to clean up the history only at the very end, once the patch is good to be landed.

**Comment by pktippa** [comment]
Hi @bajtos 

>Unfortunately you have messed up whitespace in your last commit,

That happened because of the copying the code and switching editor from notepad++ to eclipse.

>Could you please restore the original whitespace? It will make it much easier for me to comment on your changes.

I reverted the previous commit and restored original white space, please have a look at it once.

>Don't worry about it now, it's best to clean up the history only at the very end, once the patch is good to be landed.

Ok.

**Comment by bajtos** [comment]
Hi @pktippa, thank you for fixing the whitespace, the diff looks much better now!

I am afraid I am very tight on time right now and don't have enough bandwidth to finish the review. I ask for your patience, I'll get back to this as soon as my schedule permits.

**Comment by pktippa** [comment]
Hi @bajtos, 

>I am afraid I am very tight on time right now and don't have enough bandwidth to finish the review. I ask for your patience, I'll get back to this as soon as my schedule permits.

We are running a private npm using [sinopia](https://github.com/rlidwka/sinopia) and I published this fix in our private repo. Looking for this fix to go public :)

**Comment by bajtos** [comment]
@pktippa to speed things up, I made few remaining changes myself and landed the patch via 9e8a19f. Thank you for the contribution!

**Comment by pktippa** [comment]
@bajtos thanks for landing the patch. On which node module version this patch will be available.?

**Comment by bajtos** [comment]
@pktippa sorry for the delay. I have just released `loopback-component-explorer@2.3.0` that includes your patch too. Enjoy!

**Comment by pktippa** [comment]
@bajtos Thanks :+1:
